### ISSUE TITLE
Reduce dependencies

### ISSUE THREAD

**Issue by icebob** [moleculerjs/moleculer#433]
Due to [the recent event-stream vulnerability issue](https://medium.com/intrinsic/compromised-npm-package-event-stream-d47d08605502), I'm thinking to reduce the used NPM dependencies in the Moleculer core. 
Currently it is the current dependency tree (29 modules):
```
`-- moleculer@0.13.4
  +-- args@5.0.0
  | +-- camelcase@5.0.0
  | +-- chalk@2.4.1 deduped
  | +-- leven@2.1.0
  | `-- mri@1.1.1
  +-- bluebird@3.5.2
  +-- chalk@2.4.1
  | +-- ansi-styles@3.2.1
  | | `-- color-convert@1.9.3
  | |   `-- color-name@1.1.3
  | +-- escape-string-regexp@1.0.5
  | `-- supports-color@5.5.0
  |   `-- has-flag@3.0.0
  +-- es6-error@4.1.1
  +-- eventemitter2@5.0.1
  +-- fastest-validator@0.6.12
  +-- glob@7.1.3
  | +-- fs.realpath@1.0.0
  | +-- inflight@1.0.6
  | | +-- once@1.4.0 deduped
  | | `-- wrappy@1.0.2
  | +-- inherits@2.0.3
  | +-- minimatch@3.0.4
  | | `-- brace-expansion@1.1.11
  | |   +-- balanced-match@1.0.0
  | |   `-- concat-map@0.0.1
  | +-- once@1.4.0
  | | `-- wrappy@1.0.2 deduped
  | `-- path-is-absolute@1.0.1
  +-- ipaddr.js@1.8.1
  `-- lodash@4.17.11
```

There are 3 libs which have too much dependencies:
1. **args** (4 other deps)
2. **chalk** (6 other deps)
3. **glob** (12(!) other deps)

### `chalk`
- [kleur](https://github.com/lukeed/kleur) (**no deps**)

### `glob`
- [tiny-glob](https://github.com/terkelg/tiny-glob) (2 deps)

**Comment by icebob** [comment]
@lukeed do you have any awesome lib to replace the huge `glob` lib? :smile:

**Comment by lukeed** [comment]
Hey, haha, I helped work on `tiny-glob`. It's the one I use

**Comment by tinchoz49** [comment]
https://github.com/vadimdemedes/ink could be a replacement for vorpal

**Comment by lukeed** [comment]
Ink is good. I forgot to mention before that [`sade`](https://github.com/lukeed/sade) is also an option for your CLI builder. Its only dependency is `mri`, which is already in this list

**Comment by joshmanders** [comment]
Hello, we use Moleculer at my company and vorpal's vulnerabilities are a concern to us and we'd like to get this resolved. What can I do to help with this? I can fork and start a PR to replace vorpal with something else.

Thanks.

**Comment by icebob** [comment]
vorpal is more than the ink. It has a REPL mode with command & argument handling and autocomplete feature. Ink (as soon as I can see) is a react based GUI for console apps.
The better option would be to fix vulnerable packages in vorpal. I've created a [fork here](https://github.com/moleculerjs/vorpal)

**Comment by joshmanders** [comment]
Yeah I was looking over alternatives and none really seem to do what moleculer uses vorpal for.

@icebob do you plan to utilize the fork in moleculer-repl by publishing a package or directly to the github repo in the package.json? I can start working on getting the vulnerabilities patched and submit a PR to the fork.

**Comment by tinchoz49** [comment]
This is another one more close to the vorpal concept: https://github.com/drew-y/cliffy

**Comment by icebob** [comment]
@joshmanders good news. It would be good if you could fix it. I plan to publish the fixed vorpal under the org like: @moleculer/vorpal

**Comment by icebob** [comment]
@tinchoz49 great, yes, cliffy can be an alternative. _Btw, I've starred it for a while :P_

**Comment by joshmanders** [comment]
> @joshmanders good news. It would be good if you could fix it. I plan to publish the fixed vorpal under the org like: @moleculer/vorpal

Excellent. I have a PR waiting right now, just have noticed some tests are timing out so I'm trying to figure that out.

**Comment by GrosSacASac** [comment]
I heard that bluebird is no longer needed

**Comment by abdavid** [comment]
@icebob Could we add `lodash` to the list of libraries that Moleculer _could_ live without? I dissected the framework a while back while playing around with making it a first-class typescript citizen and noticed a lot of lodash usage that could be replaced by simple native js. The reason I particularly am mentioning this lib is because of the bad pratices like prototype pollution it often entails. 

https://github.com/moleculerjs/moleculer/search?q=isObject&unscoped_q=isObject

**Comment by abdavid** [comment]
Latest tree btw

```
moleculer@0.14.6 /Users/davborre/Source/moleculer
├─┬ args@5.0.1
│ ├── camelcase@5.0.0
│ ├─┬ chalk@2.4.2
│ │ ├─┬ ansi-styles@3.2.1
│ │ │ └─┬ color-convert@1.9.1
│ │ │   └── color-name@1.1.3
│ │ ├── escape-string-regexp@1.0.5
│ │ └─┬ supports-color@5.4.0
│ │   └── has-flag@3.0.0
│ ├── leven@2.1.0
│ └── mri@1.1.4
├── es6-error@4.1.1
├── eventemitter2@6.4.0
├── fastest-validator@1.4.1
├── fn-args@5.0.0
├─┬ glob@7.1.6
│ ├── fs.realpath@1.0.0
│ ├─┬ inflight@1.0.6
│ │ ├── once@1.4.0 deduped
│ │ └── wrappy@1.0.2
│ ├── inherits@2.0.3
│ ├─┬ minimatch@3.0.4
│ │ └─┬ brace-expansion@1.1.11
│ │   ├── balanced-match@1.0.0
│ │   └── concat-map@0.0.1
│ ├─┬ once@1.4.0
│ │ └── wrappy@1.0.2 deduped
│ └── path-is-absolute@1.0.1
├── ipaddr.js@1.9.1
├── kleur@3.0.3
├── lodash@4.17.15
├─┬ lru-cache@5.1.1
│ └── yallist@3.0.3
└── node-fetch@2.6.0
```

**Comment by icebob** [comment]
Yes, it would be good to drop lodash lib from the core repo. The most important method is the `_.defaultsDeep`. It is used in many places, so it's the most important to find an exact same implementation _or copy out from the lodash_.

**Comment by joshmanders** [comment]
> Yes, it would be good to drop lodash lib from the core repo. The most important method is the `_.defaultsDeep`. It is used in many places, so it's the most important to find an exact same implementation _or copy out from the lodash_.

Could just depend on `lodash.defaultsDeep` which is just the dependencies needed for that function.

**Comment by icebob** [comment]
Hmm, good idea!

**Comment by icebob** [comment]
By the way, I've just found [`cac`](https://github.com/cacjs/cac) which is same as `args` but no dependencies.

Only need to change the `glob` to avoid another [big branch](https://npm.anvaka.com/#/view/2d/moleculer)

**Comment by abdavid** [comment]
_.get used in the cacher implementation is pretty important aslo I believe. It would be great to firstly get an overview of all lodash deps to determine which could easily be switched out with vanilla js and which we need to consider keeping or replace. I’d be happy to contribute on some concrete issues like «replace usage of isObject» etc to we can gradually migrate lodash out. 

- David

**Comment by lukeed** [comment]
`cac` has dependencies (including `mri`) – it's just all inlined into one big pre-built file, which prevents any cross-dep code reuse

**Comment by icebob** [comment]
Ohh, thanks @lukeed, I didn't see it.

**Comment by lukeed** [comment]
It's actually _only_ `mri` - I thought there were more. This means `cac` basically the same as `sade`...good upgrade either way

**Comment by icebob** [comment]
@abdavid Here is a list about the used Lodash methods:
![image](https://user-images.githubusercontent.com/306521/82689665-25e5de00-9c5b-11ea-8c08-2f66e1028d70.png)

**Comment by abdavid** [comment]
Nice! I see a couple of low hanging ones. I think we can break them down pr function. 

@icebob should we maybe create an epic and materialize tasks for each function? That way we will have the oppertunity to discuss replacements contextually. Like Josh commented we could just import the complex ones.

- David

**Comment by icebob** [comment]
Ok, go ahead :)

**Comment by abdavid** [comment]
I do not have the necessary access to create anything like that. For github an epic equivalent would maybe be a project with multiple subtasks? #jiranerd

**Comment by icebob** [comment]
Here there are some native implementation: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#quick-links

**Comment by abdavid** [comment]
> Here there are some native implementation: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#quick-links

Sweet! Nice find

**Comment by abdavid** [comment]
@icebob Do we have any docs / guides on performance testing? Would be nessesary to have a before and after measurements to ensure we keep things speedy

**Comment by icebob** [comment]
Yes, you can with `npm run bench`. It executes common performance benchmark tests, so you can measure before & after with it.

**Comment by Wallacy** [comment]
@icebob Another option to replace Vorpal: https://github.com/f/omelette

Has zero dependencies;

**Comment by icebob** [comment]
Thanks @Wallacy but I think it's not a replacement of vorpal. It's just a part of it.

**Comment by AndreMaz** [comment]
One possible replacement for `vorpal` in [`moleculer-repl`](https://github.com/moleculerjs/moleculer-repl) is [`shargs`](https://github.com/Yord/shargs).

Some features:
- Highly customizable command-line arguments parser and usage documentation generator.
- 35+ opt-in features, e.g. (multiple) subcommands, spelling mistake detection, default values, and (best guess) casting.
- Synchronous and Promise-based asynchronous mode with async/await support.
- Automatic usage documentation generation with fine-grained control over layouts and styles.
- Easily extensible with your own custom parser stages and custom usage layouts.
- Extensively documented and very well tested (750+ unit and integration tests).
- Modular library layout with zero runtime dependencies.

and @Yord is currently working on [REPL mode](https://github.com/Yord/shargs-example-repl) and autocompletion

**Comment by icebob** [comment]
It sounds great. it has no dependency.

**Comment by Yord** [comment]
Hey all, shargs is meant to replace the command-line parser of another project of mine because it had way too many dependencies (27!). Seems our causes are aligned.

The REPL mode is still very fresh, but I am interested in spending more time.

**Comment by icebob** [comment]
@Yord Do you know [vorpal](https://github.com/dthree/vorpal)? Are there any other missing parts in shargs? We are using it in our [moleculer-repl](https://moleculer.services/docs/0.14/moleculer-repl.html)

**Comment by Yord** [comment]
@icebob I have never worked with vorpal.

The command-line parser parts of shargs are almost at 1.0.0, error message ergonomy is left (and bash autocomplete).

As for REPLs, shargs has basic REPL functionality based on the Node REPL and autocomplete for commands. These two parts are not tested, yet.

I could imagine shargs REPL misses text colors and formatting (aka tables) unless the action output does that.

Is there anything apparent to vorpal users that I miss?

**Comment by AndreMaz** [comment]
> I could imagine shargs REPL misses text colors and formatting (aka tables) unless the action output does that.

You are right, the coloring and formatting are done directly in action (e.g., [`service`](https://github.com/moleculerjs/moleculer-repl/blob/master/src/commands/services.js) )

> Edited to add: Shargs REPL misses pipes and prompts.

I think that we don't use either of those features. We only use "a small portion" of vorpal

**Comment by icebob** [comment]
Would be good to create a branch in moleculer-repl and trying to add Shargs instead of vorpal. @AndreMaz do you have some capacity to work on it? Maybe @Yord can help if something is missing :)

**Comment by Yord** [comment]
@icebob, @AndreMaz I will gladly ship in if I can be of help.

**Comment by AndreMaz** [comment]
@icebob I look at it during weekend. In theory it shouldn't be very complex. I just need to find the equivalent for vorpal's syntax. E.g.:

```js
vorpal
  .command('foo <requiredArg> [optionalArg]')
  .autocomplete({
    data() {
        return // Return autocomplete array
  })
  .option('-v, --verbose', 'Print foobar instead.')
  .description('Outputs "bar".')
  .alias('foosball')
  .action(function(args, callback) {
    if (args.options.verbose) {
      this.log('foobar');
    } else {
      this.log('bar');
    }
    callback();
  });
```

@Yord do you have any tutorial for newbies? Your docs are very good but I didn't found any introductory tutorial.

**Comment by Yord** [comment]
@AndreMaz Fortunately I wrote one just yesterday:

[https://github.com/Yord/shargs-tutorial-git/blob/master/README.md](https://github.com/Yord/shargs-tutorial-git/blob/master/README.md)

**Comment by AndreMaz** [comment]
Sweet. Thanks

**Comment by Yord** [comment]
@AndreMaz Just wanted to let you know I have incorporated recent updates into [shargs-example-repl](https://github.com/Yord/shargs-example-repl).

**Comment by AndreMaz** [comment]
@Yord thanks for the info. Can you add the changelog to your repos? It helps to track the changes

**Comment by Yord** [comment]
@AndreMaz Yes, I will do that. Sorry for the sloppy work. I usually start a changelog with the first stable version, but here, it definitely makes sense to start right now.

Changelogs:
+    *[shargs-example-repl 0.4.0](https://github.com/Yord/shargs-example-repl/releases/tag/0.4.0)*
+    [shargs-repl 0.2.0](https://github.com/Yord/shargs-repl/releases/tag/0.2.0)
+    [shargs 0.24.6](https://github.com/Yord/shargs/releases/tag/0.24.6)

**Comment by gautaz** [comment]
Hello, sorry to barge in but I think that what I raised [here](https://github.com/moleculerjs/moleculer/issues/1169#issuecomment-1381849406) might be related to this issue.

My main concern is related to the fact that some of us (i have no idea how many) are using the Moleculer broker without the runner.
I was wondering if moving the broker to something like `moleculer-core` (or the runner to `moleculer-runner`) would be an acceptable solution.

The runner is great but it also comes with its cost in term of dependencies and I guess that part of the dependencies raised in this issue is in fact due to the runner (another batch is certainly due to the REPL).
:warning: This last one is an assumption, I did no check the source code so I may be totally wrong (which would be far from being the first time :laughing:).

**Comment by icebob** [comment]
I plan to split this repo to multiple packages in v1.0 and release them under `@moleculer/....`
